### PR TITLE
Feat: history api request refactoring

### PR DESCRIPTION
**Pull request by Victor-Salomon**
This PR contains a refactor of the history API route from a GET request to a POST request. 

Issue related: [here](https://linear.app/velocitylabs/issue/VEL-58/refactoring-apihistory-route-from-a-get-request-to-a-post-request).

It includes 
- the appropriate types
- the Zod parse schema to validate the expected body

I'd like to confirm one concern with you: Currently, we fetch the last two weeks of transfers, expecting to find the user's ongoing transfers within the result. Should we consider reducing this two-week buffer to a few days only in production? I do not see any transfer pending for more than a few days. 

What do you think @NunoAlexandre

### CODE REVIEW

**Comment by NunoAlexandre** [comment]
> I'd like to confirm one concern with you: Currently, we fetch the last two weeks of transfers, expecting to find the user's ongoing transfers within the result. Should we consider reducing this two-week buffer to a few days only in production? I do not see any transfer pending for more than a few days.
> 
> What do you think @NunoAlexandre

@Victor-Salomon ideally we wouldn't need to "search through" anything but instead we would have a transfer hash and look up the status of that directly. What's in the way of making that possible?

If we are indeed limited to that approach, I'd say 1 day is a good safe guard. Users might only get back to the app a day later and it's important that Turtle gives continuation to the tx tracking  👍

**Comment by Victor-Salomon** [comment]
@NunoAlexandre, it seems that after you merged the release/v1 branch in mine, the build error I mentioned the other day is still there. You can see it on the release/v1 branch itself, the last build failed. 
Let me know if you want me to handle it ;)

**Comment by NunoAlexandre** [comment]
Sorry @Victor-Salomon, I did that by instinct. I will fix the build error 👍

**Comment by Victor-Salomon** [comment]
No worries, the PR is still open as a draft anyway ;)

**Comment by NunoAlexandre** [comment]
> No worries, the PR is still open as a draft anyway ;)

Fixed ;)

**Comment by Victor-Salomon** [comment]
> > I'd like to confirm one concern with you: Currently, we fetch the last two weeks of transfers, expecting to find the user's ongoing transfers within the result. Should we consider reducing this two-week buffer to a few days only in production? I do not see any transfer pending for more than a few days.
> > What do you think @NunoAlexandre
> 
> @Victor-Salomon ideally we wouldn't need to "search through" anything but instead we would have a transfer hash and look up the status of that directly. What's in the way of making that possible?
> 
> If we are indeed limited to that approach, I'd say 1 day is a good safe guard. Users might only get back to the app a day later and it's important that Turtle gives continuation to the tx tracking 👍

The new approach you mentioned is the one I started to implement regarding the tracking of transfers from Parachains to ETH. However, we will still be limited to the current approach in some cases. We can aim to progressively migrate to the new approach as it requires more work and is out of the scope of this PR. 

I will decrease the two-week fetch to 2 days as you mentioned. (the current ongoing day + an extra safeguard day). I agree with this.

**Comment by NunoAlexandre** [comment]
> > > I'd like to confirm one concern with you: Currently, we fetch the last two weeks of transfers, expecting to find the user's ongoing transfers within the result. Should we consider reducing this two-week buffer to a few days only in production? I do not see any transfer pending for more than a few days.
> > > What do you think @NunoAlexandre
> > 
> > 
> > @Victor-Salomon ideally we wouldn't need to "search through" anything but instead we would have a transfer hash and look up the status of that directly. What's in the way of making that possible?
> > If we are indeed limited to that approach, I'd say 1 day is a good safe guard. Users might only get back to the app a day later and it's important that Turtle gives continuation to the tx tracking 👍
> 
> @NunoAlexandre The new approach you mentioned is the one I started to implement regarding the tracking of transfers from Parachains to ETH. However, we will still be limited to the current approach in some cases. We can aim to progressively migrate to the new approach as it requires more work and is out of the scope of this PR.
> 
> I will decrease the two-week fetch to 2 days as you mentioned. (the current ongoing day + an extra safeguard day). I agree with this.

Sounds good :)

File: app/src/app/api/history/route.ts [hunk 1]
```diff
@@ -3,26 +3,38 @@ export const dynamic = 'force-dynamic' // Always run dynamically
...
+        toEthereum: [],
+      }
+
+      ongoingTransfers.map(transfer => {
+        if (transfer.direction === Direction.ToEthereum) {
+          transfers.toEthereum.push(transfer)
+        } else {
+          transfers.toPolkadot.push(transfer)
+        }
+        return transfer
```
[the remark this answers is not in the document]
Comment (Victor-Salomon) [reply app/src/app/api/history/route.ts:34]: ✅ I will update as it makes sense. 
I knew this one was 'temporary' like this, as after the new tracking I will have to differentiate toEthereum transfer when coming from Assethub or a Parachain.

File: app/src/app/api/history/route.ts [hunk 2]
```diff
@@ -35,10 +47,21 @@ const getCachedTransferHistory = unstable_cache(
...
+    // Returns 400 if body does not respect the expected schema
+    if (!requestValue.success) {
+      return NextResponse.json({ error: requestValue.error }, { status: 400 })
+    }
+    const { ongoingTransfers } = requestValue.data
+    // Returns 200 if ongoingTransfers is empty
+    if (!ongoingTransfers.length) return NextResponse.json([], { status: 200 })
+
+    const history = await getCachedTransferHistory(ongoingTransfers)
+    return NextResponse.json(history, { status: 200 })
```
[the remark this answers is not in the document]
Comment (Victor-Salomon) [reply app/src/app/api/history/route.ts:64]: We discussed this implementation with Noah before he started. 
Triggering this `getCachedTransferHistory` in the API/server side, allows us to handle some cache. Here it's cached globally, which is better than a client-side cache solution that would only cache data for one user.

File: app/src/models/transfer.ts [hunk 3]
```diff
@@ -5,26 +5,38 @@ import { Environment } from '@/store/environmentStore'
...
-
   // Contextual
   environment: Environment // to access context
   // TODO(nuno): we can have multiple types of transfer and have this depend on that type.
   // that way we can support different fields, for example for xcm-only transfers in the future.
   sendResult?: toEthereum.SendResult | toPolkadot.SendResult
 }
 
+export interface OngoingTransferWithDirection extends RawTransfer {
+  direction: Direction
```
[the remark this answers is not in the document]
Comment (Victor-Salomon) [reply app/src/models/transfer.ts:32]: In the tracking, I can't pass the `StoredTransfer[]` object without serializing its `BN` values, which won't be used. I've created a `formattedTransfers` object that stores the `Direction`, so I don't have to compute it every time I pass the ongoing transfers as props.

With this new typing implementation, we now use a common base type, `RawTransfer`, which gets extended based on the app's type needed. For example, StoredTransfer doesn't need to include the Direction.

If you don't like the approach, I can revert it. However, I'll still need to create my own formattedType, which can be done using utilities like Pick<Type, Keys> or Omit<Type, Keys>.

[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply app/src/models/transfer.ts:32]: I understand. That makes sense, I like how you thought about this :)

**Review by NunoAlexandre — APPROVED** [review]
Trés beau 🤩

**Pull request merged**